File: List/list.py

class List:
    """ Dynamic list implementation class. """

    def __init__(self):
        self._size = 0
        self._capacity = 1
        self._items = [None for _ in range(self._capacity)]
    
    def add(self, item):
        self._size += 1
        if self._size >= self._capacity: self._capacity *= 2
        if self._capacity > len(self._items):
            tmp = [None for _ in range(self._capacity)]
            for i in range(self._size - 1):
                tmp[i] = self._items[i]
            self._items = tmp
        self._items[self._size - 1] = item
    
    def display(self):
        for i in range(self._size):
            print(self._items[i], end=' ')
        print()
    
    def size(self):
        return self._size
    
    def get(self, idx):
        return self._items[idx] if idx <= self._size else None
    
    def indexOf(self, item):
        for i in range(self._size):
            if item == self._items[i]: return i
        
        return -1
    
    def remove(self, idx=None, item=None):
        if idx is None and item is not None:
            idx = self.indexOf(item) 
        if 0 <= idx <= self._size:
            for i in range(idx, self._size - 1):
                self._items[i] = self._items[i + 1]
            self._size -= 1
            self._items[self._size] = self._items[self._capacity - 1]

    def isEmpty(self):
        return self._size == 0
    
    def contains(self, item):
        return self.indexOf(item) != -1
    
    def first(self):
        return self._items[0] if not self.isEmpty() else None
    
    def last(self):
        return self._items[self._size - 1] if not self.isEmpty() else None

    def set_(self, idx, item):
        if 0 <= idx < self._size: self._items[idx] = item

    def capacity(self):
        return self._capacity

File: List/list.js

class List {

  constructor() {
    this._size = 0
    this._capacity = 1
    this.items = [0]
  }

  add(item) {
    if (++this._size > this._capacity) {
      const tmp = new Array(this._capacity)
      for (let i = 0; i < this._size - 1; i++) {
        tmp[i] = this.items[i]
      }

      this.items = tmp;
    }

    this.items[this._size - 1] = item;
  }

  display() {
    return `[ ${this.items.slice(0, this._size).join(', ')} ]`
  }

  size() {
    return this._size
  }

  get(idx) {
    return this.checkIndex(idx) 
      ? this.items[idx]
      : undefined
  }

  indexOf(item) {
    for (let i = 0; i < this._size; i++) {
      if (item === this.items[i]) {
        return i
      }
    }

    return -1
  }

  removeByIndex(idx) {
    if (this.checkIndex(idx)) {
      for (let i = idx; i < this._size; i++) {
        this.items[i] = this.items[i + 1]
      }

      this.items[--this._size] = this.items[this._capacity - 1]
    }
  }

  remove(item) {
    const idx = this.indexOf(item)
    if (idx !== -1) {
      this.removeByIndex(idx)
    }
  }

  isEmpty() {
    return this._size === 0
  }

  contains(item) {
    return this.indexOf(item) !== -1
  }

  first() {
    return this.isEmpty()
      ? undefined
      : this.items[0]
  }

  last() {
    return this.isEmpty()
      ? undefined
      : this.items[this._size - 1]
  }

  set(idx, item) {
    if (this.checkIndex(idx)) {
      this.items[idx] = item
    }
  }

  capacity() {
    return this._capacity
  }

  checkIndex(idx) {
    return idx >= 0 && idx <= this._size
  }
}
